File: parse_coverage.py
import os
import re
from bs4 import BeautifulSoup

def get_coverage_from_html(html_path, base_dir=""):
    with open(html_path, 'r', encoding='utf-8') as f:
        soup = BeautifulSoup(f, 'html.parser')
    
    rows = soup.find_all('tr')[1:] # Skip header
    results = []
    
    for row in rows:
        file_td = row.find('td', class_='file')
        if not file_td:
            continue
        
        file_name = file_td.get('data-value')
        
        # Branch coverage is the 3rd <td> with data-value or check pct class
        tds = row.find_all('td')
        # Based on index.html: 
        # 0: File, 1: Pic, 2: Statements Pct, 3: Statements Abs, 
        # 4: Branches Pct, 5: Branches Abs, 6: Functions Pct, ...
        
        try:
            branch_pct = float(tds[4].get('data-value'))
        except (IndexError, ValueError, TypeError):
            branch_pct = 100.0
            
        full_path = os.path.join(base_dir, file_name).replace('\\', '/')
        results.append((full_path, branch_pct))
    return results

all_results = []
# Start with root
all_results.extend(get_coverage_from_html('coverage/ngt-template/index.html'))

# Crawl subdirectories
for root, dirs, files in os.walk('coverage/ngt-template'):
    if 'index.html' in files:
        rel_path = os.path.relpath(root, 'coverage/ngt-template')
        if rel_path == '.':
            continue
        all_results.extend(get_coverage_from_html(os.path.join(root, 'index.html'), rel_path))

# Filter out directories (they end with /index.html in the results usually, but we want files)
# In Istanbul/nyc, index.html lists files and directories. 
# Files usually end in .ts or .js or similar. Directories don't.
final_files = [res for res in all_results if res[0].endswith(('.ts', '.js', '.html', '.scss'))]

# Sort by branch coverage
final_files.sort(key=lambda x: (not x[0].startswith('app/page/game'), x[1]))

# Output top 15
for path, pct in final_files[:15]:
    print(f"{path}: {pct}%")

File: parse_coverage.js
const fs = require('fs');
const path = require('path');

function getCoverageFromHtml(htmlPath, baseDir = '') {
    const content = fs.readFileSync(htmlPath, 'utf8');
    const results = [];
    
    // Simple regex to match rows. 
    // This is fragile but often works for Istanbul HTML output.
    const rowRegex = /<tr[^>]*>[\s\S]*?<\/tr>/g;
    const matches = content.match(rowRegex);
    if (!matches) return [];

    // Skip header
    for (let i = 1; i < matches.length; i++) {
        const row = matches[i];
        const fileMatch = row.match(/data-value="([^"]+)"/);
        if (!fileMatch) continue;
        
        const fileName = fileMatch[1];
        
        // Find all data-value attributes in the row
        const dataValues = [...row.matchAll(/data-value="([^"]+)"/g)].map(m => m[1]);
        
        // Columns: 0: File, 1: Pic, 2: Statements Pct, 3: Statements Abs, 4: Branches Pct
        const branchPct = parseFloat(dataValues[4]) || 0;
        
        const fullPath = path.join(baseDir, fileName).replace(/\\/g, '/');
        results.push({ path: fullPath, pct: branchPct });
    }
    return results;
}

function walkDir(dir, base = '') {
    let results = [];
    const list = fs.readdirSync(dir);
    
    if (list.includes('index.html')) {
        results = results.concat(getCoverageFromHtml(path.join(dir, 'index.html'), base));
    }

    list.forEach(file => {
        const fullPath = path.join(dir, file);
        const stat = fs.statSync(fullPath);
        if (stat && stat.isDirectory()) {
            results = results.concat(walkDir(fullPath, path.join(base, file)));
        }
    });

    return results;
}

const allResults = walkDir('coverage/ngt-template');

// Filter for files only (Istanbul uses data-value for the file/dir name)
// In the index.html, it lists files in the current dir and subdirs.
// We want to avoid duplicates if possible, and only keep actual source files.
const filesOnly = allResults.filter(r => r.path.match(/\.(ts|js|html|scss)$/));

// Sort: prioritize app/page/game, then by branch coverage ascending
filesOnly.sort((a, b) => {
    const aInGame = a.path.startsWith('app/page/game');
    const bInGame = b.path.startsWith('app/page/game');
    
    if (aInGame && !bInGame) return -1;
    if (!aInGame && bInGame) return 1;
    
    return a.pct - b.pct;
});

// Dedup (Istanbul might list them in multiple index.html files)
const seen = new Set();
const uniqueFiles = [];
for (const f of filesOnly) {
    if (!seen.has(f.path)) {
        seen.add(f.path);
        uniqueFiles.push(f);
    }
}

uniqueFiles.slice(0, 15).forEach(f => {
    console.log(`${f.path}: ${f.pct}%`);
});
